### PR TITLE
Add depth module

### PR DESCRIPTION
**Pull request by ejm714**
This implementation takes in a folder of videos (or csv with video filepaths) and outputs a csv of distance estimates for each frame (and detection). 

For frames with no detection, distance is null. If there are multiple detections in that frame, there is one line per detection. For example

```
filepath,time,distance
tests/assets/videos/data/raw/goualougo_2013/gorillas_2013/MPI_FID_20_Duboscia/14-Jun-2013/FID_20_Duboscia_2013-6-14_0083.AVI,0,7.4
tests/assets/videos/data/raw/goualougo_2013/gorillas_2013/MPI_FID_20_Duboscia/14-Jun-2013/FID_20_Duboscia_2013-6-14_0083.AVI,0,7.4
tests/assets/videos/data/raw/goualougo_2013/gorillas_2013/MPI_FID_20_Duboscia/14-Jun-2013/FID_20_Duboscia_2013-6-14_0083.AVI,1,
tests/assets/videos/data/raw/goualougo_2013/gorillas_2013/MPI_FID_20_Duboscia/14-Jun-2013/FID_20_Duboscia_2013-6-14_0083.AVI,2,
tests/assets/videos/data/raw/goualougo_2013/gorillas_2013/MPI_FID_20_Duboscia/14-Jun-2013/FID_20_Duboscia_2013-6-14_0083.AVI,3,
```

Not included in this PR:
- Masking out other detections where there is more than 1. Under the current implementation, if there are two detections in the frame, the distance estimate will be the same for both detections (since the same input is used for both). This allows us to capture the number of detections in the frame, but not differential distances.

Notes:
- In the original implementation, there was a bug where the order of the first two frames were swapped: [-1, -2, 0, 1, 2]. This bug is maintained to match what the model expects

Closes https://github.com/drivendataorg/pjmf-zamba/issues/57

Remaining to do:
- [x] update docs home page to include depth model (and update CLI output)

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -66,7 +66,7 @@ densepose =
     detectron2-densepose @ git+https://github.com/facebookresearch/detectron2@main#subdirectory=projects/DensePose
 
 [flake8]
-ignore = E203, E501, W503
+ignore = E203, E501, W503, E722
```
Comment (ejm714) [inline setup.cfg:69]: E722 allows bare excepts

File: zamba/models/depth_estimation/depth_manager.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        det_metadata = self.detection_dict[detection_idx]
+        det_frame = det_metadata["frame"]
+        det_video = det_metadata["video"]
+
+        # set up input array of frames within window of detection
+        # frames are stacked channel-wise
+        input = np.zeros((self.height, self.width, self.channels * self.num_frames))
+        n = 0
+        for frame_idx in range(det_frame - self.window_size, det_frame + self.window_size + 1):
+            input[:, :, n : n + self.channels] = self.cached_frames[det_video][
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:123]: worth looking into ways to be more efficient in the memory management here (e.g. no allocating ahead and doing the stack)

File: zamba/models/depth_estimation/depth_manager.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                f"frame_{frame_idx}"
+            ]
+            n += self.channels
+
+        # normalize and convert to tensor
+        input = normalize(input)
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:129]: let's do this before we put the frame in the cached frames dict so we only have to normalize and transpose once

File: zamba/models/depth_estimation/depth_manager.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+        Args:
+            model_cache_dir (Path): Path for downloading and saving model weights.
+            weight_download_region (str): s3 region to download pretrained weights from.
+                Options are "us" (United States), "eu" (Europe), or "asia" (Asia Pacific).
+                Defaults to "us".
+            batch_size (int, optional): Batch size to use for inference. Defaults to 64.
+                Note: a batch is a set of frames, not videos, for the depth model.
+            tta (int, optional): Number of flips to apply for test time augmentation.
+            use_log (bool, optional): Whether to take the exponential of the predictions
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:155]: we may want to remove this as long as we're outputting probabilities

File: zamba/models/depth_estimation/depth_manager.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        self.model_weights = model_cache_dir / model["weights"]
+        if not self.model_weights.exists():
+            model_cache_dir.mkdir(parents=True, exist_ok=True)
+            self.model_weights = download_weights(
+                model["weights"], model_cache_dir, weight_download_region
+            )
+
+        # automatically use CPU if no cuda available
+        if not torch.cuda.is_available():
+            self.device = "cpu"
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:173]: we should expose `device` this further up (maybe call gpus for consistency)

File: zamba/models/depth_estimation/depth_manager.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+        # load dataset
+        test_dataset = DepthDataset(filepaths)
+        test_loader = torch.utils.data.DataLoader(
+            test_dataset,
+            batch_size=self.batch_size,
+            shuffle=False,
+            sampler=None,
+            collate_fn=None,
+            num_workers=min(self.batch_size, 8),
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:191]: let's expose this

File: zamba/models/depth_estimation/depth_manager.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        lengths = [np.arange(test_dataset.cached_frames[v]["video_length"]) for v in videos]
+
+        # create one row per frame for duration of video
+        output = pd.Series(index=videos, data=lengths).explode().to_frame().reset_index()
+        output.columns = ["filepath", "time"]
+
+        # merge in predictions
+        if len(predictions) > 0:
+            output = output.merge(predictions, on=["filepath", "time"], how="outer")
+        return output
```
Comment (ejm714) [inline zamba/models/depth_estimation/depth_manager.py:242]: if no predictions (len(preds) == 0), need to add distance column (nan)

File: tests/test_depth.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    config = DepthEstimationConfig(
+        filepaths=pd.DataFrame(columns=["filepath"], data=[invalid_video]),
+        save_to=tmp_path / "preds.csv",
+    )
+    config.run_model()
+
+    # ensure outputs get written out and but is empty since video could not be loaded
+    preds = pd.read_csv(tmp_path / "preds.csv")
+    assert len(preds) == 0
```
Comment (ejm714) [inline tests/test_depth.py:81]: check that i have all three columns

File: tests/test_depth.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    filepaths = pd.read_csv(two_video_filepaths).filepath.values
+    preds = dem.predict(filepaths)
+
+    assert preds.shape == (77, 3)
+    assert preds.distance.notnull().sum() == 2
+    assert preds.filepath.nunique() == 2
+
+    # two detections for frame 0
+    assert preds.set_index(["filepath", "time"]).loc[filepaths[0], 0].shape[0] == 2
+
```
Comment (ejm714) [inline tests/test_depth.py:41]: add a check that the distance estimates are the same as what we get through running through the comp winning model (use a video with single animal)

**Review by ejm714 — COMMENTED** [review]
Comments from synchronous code review

File: zamba/models/depth_estimation/depth_manager.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                # if anything is detected in the frame, save out relevant frames
+                if len(detections) > 0:
+                    logger.debug(f"{len(detections)} detection(s) found at second {frame_idx}.")
+
+                    # get frame indices around frame with detection
+                    min_frame = frame_idx - self.window_size
+                    max_frame = frame_idx + self.window_size
+
+                    # add relevant resized frames to dict if not already added
+                    # if index is before start or after end of video, use an array of zeros
```
[the remark this answers is not in the document]
Comment (ejm714) [reply zamba/models/depth_estimation/depth_manager.py:73]: comp code uses zeros: https://github.com/drivendataorg/deep-chimpact-winners/blob/fce53504040ecfa1fe5627bb2bd142ae7696a93a/2nd%20Place/test.py#L117

File: docs/docs/models/depth.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
+# Depth estimation
+
+## Background
```
Comment (pjbull) [inline docs/docs/models/depth.md:3]: Can we start with a sentence like "Our depth estimation model is used to estimate the distance an animal is from the camera, which is an input into model used to estimate animal abundance."

**Review by pjbull — COMMENTED** [review]
Docs seem great—just one thought on a clarifying intro sentence!

File: docs/docs/models/depth.md (hunk 11 above)
Comment (ejm714) [reply docs/docs/models/depth.md:3]: good thought, thanks

File: zamba/models/depth_estimation/depth_manager.py (hunk 3 above)
Comment (ejm714) [reply zamba/models/depth_estimation/depth_manager.py:129]: not doing as it causes a `[W NNPACK.cpp:53] Could not initialize NNPACK! Reason: Unsupported hardware.` error. the current implementation mimics the comp winning code so this seems reasonable enough

File: zamba/models/depth_estimation/depth_manager.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+                    for i in range(min_frame, max_frame + 1):
+                        if f"frame_{i}" not in cached_frames[video_filepath].keys():
+                            try:
+                                selected_frame = np.array(
+                                    # PIL expects size to be (w, h)
+                                    Image.fromarray(arr[i]).resize((self.width, self.height))
```
[the remark this answers is not in the document]
Comment (ejm714) [reply zamba/models/depth_estimation/depth_manager.py:81]: done

**Comment by ejm714** [comment]
All PR comments have been addressed. Checked memory usage with scalene and confirmed our predictions are within 0.1 of the competition code, often times closer to a difference of 0.03. This is likely because the competition code does some writing out and reading in with open cv and also uses albumentations to resize (which uses open cv under the hood). We use torchvision transforms instead. Distances are usually measured in increments of 0.5 so these differences don't seem problematic.

Tests are passing locally so I need to look into why they're failing on github.

@pjbull if you have a moment to give the diff a quick read as a backstop, that'd be great. Here's the [link](https://github.com/drivendataorg/zamba/pull/247/files/e1cef48bbdfa046067a1c28b1af935543646adcd..7bcc48afa4c44d796ca557376252cc45d20bc2ff#)
